File: BaseNumberConverter.c
/*
 * Base number converter
 * Convert decimal to binary/octal/hexadecimal number
 * 
 * Author: Mohamamd Najmi Bin Bachok
 * Version: 1.0.0
 */

#include <stdlib.h>
#include <stdio.h>

const char* basenumber(unsigned int, unsigned int, const char*);
const char* binary(unsigned int);
const char* octal(unsigned int);
const char* hexadecimal(unsigned int);

int main(int argc, char** argv)
{
	printf_s("1 = bin(0x%s)\n", binary(1));
	printf_s("1 = oct(0x%s)\n", oct(1));
	printf_s("1 = hex(0x%s)", hex(1));
	return 0;
}

/*!
 @function basenumber
 @param decimal: The decimal number to be converted
 @param base: The base of the number
 @param numbers: The numbers in the base number given
 @result A number in the specified base of the given decimal number
*/
const char* basenumber(unsigned int decimal, unsigned int base, const char* numbers)
{
	unsigned int length = 0;
	unsigned int decimal_copy = decimal;

	while (decimal > 0)
	{
		decimal = decimal / base;
		length++;
	}

	char* outnumber = (char*)calloc(length, sizeof(char));

	if (outnumber == NULL)
	{
		printf_s("Cannot allocate memory");
		return '\0';
	}

	length--;

	while (decimal_copy > 0)
	{
		*(outnumber + length) = numbers[decimal_copy % base];
		decimal_copy = decimal_copy / base;
		length--;
	}

	return outnumber;
}

/*!
 @function binary
 @param decimal: The decimal number to be converted
 @result A binary number of the given decimal number
*/
const char* binary(unsigned int decimal)
{
	return basenumber(decimal, 2, "01");
}

/*!
 @function octal
 @param decimal: The decimal number to be converted
 @result A octal number of the given decimal number
*/
const char* octal(unsigned int decimal)
{
	return basenumber(decimal, 8, "01234567");
}

/*!
 @function hexadecimal
 @param decimal: The decimal number to be converted
 @result A hexadecimal number of the given decimal number
*/
const char* hexadecimal(unsigned int decimal)
{
	return basenumber(decimal, 16, "0123456789ABCDEF");
}

File: Python/basenumberconverter.py
"""
# Generate decimal to binary/octal/hexadecimal number
# Author: Mohammad Najmi
# Version: 1.0.0
"""

def basenumberconverter(dec, base, numbers):
    """
    Convert decimal number to number of given base

    Parameters:
    dec (int)  : Decimal to convert
    base (int) : Base of the number to convert into
    numbers (str): Numbers of the given base

    Returns:
    str: The converted number in given base
    """

    length = 0
    dec_copy = dec
    outnumber = ""

    while dec > 0:
        dec = dec // base
        length = length + 1

    while dec_copy > 0:
        outnumber = outnumber + numbers[dec_copy % base]
        dec_copy = dec_copy // base
        length = length - 1

    return outnumber[::-1]

def binary(dec):
    """
    Convert decimal number to binary

    Parameters:
    dec (int) : Decimal to convert

    Returns:
    str: The binary number of the given decimal
    """
    return basenumberconverter(dec, 2, "01")

def octal(dec):
    """
    Convert decimal number to octal

    Parameters:
    dec (int) : Decimal to convert

    Returns:
    str: The octal number of the given decimal
    """
    return basenumberconverter(dec, 8, "01234567")

def hexadecimal(dec):
    """
    Convert decimal number to hexadecimal

    Parameters:
    dec (int) : Decimal to convert

    Returns:
    str: The hexadecimal number of the given decimal
    """
    return "0x" + basenumberconverter(dec, 16, "0123456789ABCDEF")
